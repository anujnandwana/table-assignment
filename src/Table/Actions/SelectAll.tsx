import React from "react";
import { TableData } from "../../data/data";

interface SelectAllProps {
    selectAll: boolean;
    setSelectedRowIds: (ids: string[]) => void;
    tableData: TableData[]
}

const SelectAll: React.FC<SelectAllProps> = ({ selectAll, setSelectedRowIds, tableData }) => {

   /**
   * Handles the selection and deselection of selectAll using keyboard and mouse
   * @param isSelected - The selectAll checkbox is selected or not
   * @param keyType - The type of key pressed
   */
    const handleSelectAll = function (isSelected: boolean, keyType: string | undefined) {
        
        if (keyType !== undefined && keyType !== "Enter") {
            return;
        }

        if ((isSelected && keyType === undefined) || (keyType === "Enter" && isSelected)) {
            setSelectedRowIds( tableData.map((row) => row.id as string));
        } else {
            setSelectedRowIds([]);
        }
    };

    return (
        <input
            type="checkbox"
            aria-label="Select All Rows"
            checked={selectAll}
            onChange={(event) => handleSelectAll(event.target.checked, undefined)}
            onKeyUp={(event) => handleSelectAll(!selectAll, event.key)}
        />
    );
};

export default SelectAll;
