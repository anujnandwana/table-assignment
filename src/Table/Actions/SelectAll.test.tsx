import SelectAll from './SelectAll';
import { mount } from 'enzyme';
import { TableData } from '../../data/data';

describe('SelectAll component', () => {
  const mockData: TableData[] = [
    { id: "mockedId", device: 'Device 1', path: '/path/1', name: 'jui', status: 'available' },
    { id: "mockedId1", device: 'Device 2', path: '/path/2', name: 'jui2', status: 'available' }
  ];

  const setSelectedRowIds = jest.fn()
  const tableData = mockData;

  test('DownloadAction component should match snapshot', () => {
    const component = mount(<SelectAll selectAll={false} setSelectedRowIds={setSelectedRowIds} tableData = {tableData}/>);
    expect(component).toMatchSnapshot();
  });

  test('selectAll Checkbox on click select all rows when checked and remove all rows when unchecked', () => {
    const setSelectedRowIds = jest.fn()
    const component = mount(<SelectAll selectAll={false} setSelectedRowIds={setSelectedRowIds} tableData = {tableData}/>);
    const selectAllCheckbox = component.find("input");

    selectAllCheckbox.at(0).simulate('change', { target: { checked: true } });

    expect(selectAllCheckbox.length).toBe(1);
    expect(setSelectedRowIds).toHaveBeenCalledWith(["mockedId","mockedId1"])

    selectAllCheckbox.at(0).simulate('change', { target: { checked: false } });

    expect(setSelectedRowIds).toHaveBeenLastCalledWith([])
  });

  test('selectAll Checkbox on key Up select all rows when checked', () => {
    const setSelectedRowIds = jest.fn()
    const component = mount(<SelectAll selectAll={false} setSelectedRowIds={setSelectedRowIds} tableData = {tableData}/>);
    const selectAllCheckbox = component.find("input");

    selectAllCheckbox.at(0).simulate('keyup', { key: 'Enter', keyCode: 13, which: 13 });

    expect(selectAllCheckbox.length).toBe(1);
    expect(setSelectedRowIds).toHaveBeenCalledWith(["mockedId","mockedId1"]);
  });
  
});
