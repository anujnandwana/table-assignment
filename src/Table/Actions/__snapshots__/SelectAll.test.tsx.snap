// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectAll component DownloadAction component should match snapshot 1`] = `ReactWrapper {}`;
